packages/@aws-cdk/pipelines/lib/docker-registry.ts

```diff
@@ -0,0 +1,164 @@
+    return new ExternalDockerRegistry('index.docker.io', opts);
+  }
+
+  /** Docstrings */
+  public static fromCustomRegistry(registryDomain: string, opts: ExternalDockerRegistryOptions): DockerRegistry {
+    return new ExternalDockerRegistry(registryDomain, opts);
+  }
+
+  /** Docstrings */
+  public static fromEcr(repositories: ecr.IRepository[], opts?: EcrDockerRegistryOptions): DockerRegistry {
+    return new EcrDockerRegistry(repositories, opts ?? {});
+  }
+
+  /** Docstrings */
+  public abstract registryDomain: string;
+  /** Docstrings */
+  public abstract grantRead(grantee: iam.IGrantable): void;
+  /**
+   * Docstrings
+   * @internal
+   */
+  public abstract _renderCdkAssetsConfig(): DockerRegistryCredentialSource
+}
+
+/** Docstrings */
+export interface ExternalDockerRegistryOptions {
+  /** Docstrings */
+  readonly secret: secretsmanager.ISecret;
+  /** Docstrings */
+  readonly secretUsernameField?: string;
+  /** Docstrings */
+  readonly secretPasswordField?: string;
+  /** Docstrings */
+  readonly assumeRole?: iam.IRole
+  /** Docstrings */
+  readonly usages?: DockerRegistryUsage[];
+}
+
+/** Docstrings */
+export interface EcrDockerRegistryOptions {
+  /** Docstrings */
+  readonly assumeRole?: iam.IRole
+  /** Docstrings */
+  readonly usages?: DockerRegistryUsage[];
+}
+
+/** Docstrings */
+export enum DockerRegistryUsage {
+  SYNTH,
+  SELF_UPDATE,
+  ASSET_PUBLISHING,
+};
+
+/** Docstrings */
+class ExternalDockerRegistry extends DockerRegistry {
+  constructor(public readonly registryDomain: string, private readonly opts: ExternalDockerRegistryOptions) {
+    super();
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
+    if (this.opts.assumeRole) {
+      grantee.grantPrincipal.addToPrincipalPolicy(new iam.PolicyStatement({
+        actions: ['sts:AssumeRole'],
+        resources: [this.opts.assumeRole.roleArn],
+      }));
+    }
+    const role = this.opts.assumeRole ?? grantee;
+    this.opts.secret.grantRead(role);
+  }
+
+  /**
+   * Docstrings
+   * @internal
+   */
+  public _renderCdkAssetsConfig(): DockerRegistryCredentialSource {
+    return {
+      secretsManagerSecretId: this.opts.secret.secretArn,
+      secretsUsernameField: this.opts.secretUsernameField,
+      secretsPasswordField: this.opts.secretPasswordField,
+      assumeRoleArn: this.opts.assumeRole?.roleArn,
+    };
+  }
+}
+
+/** Docstrings */
+class EcrDockerRegistry extends DockerRegistry {
+  public readonly registryDomain: string;
+
+  constructor(private readonly repositories: ecr.IRepository[], private readonly opts: EcrDockerRegistryOptions) {
+    super();
+
+    if (repositories.length === 0) {
+      throw new Error('must supply at least one `ecr.IRepository` to create an `EcrDockerRegistry`');
+    }
+    this.registryDomain = Fn.select(0, Fn.split('/', repositories[0].repositoryUri));
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
+    if (this.opts.assumeRole) {
```

Comment:
Why don't we have `Role#grantAssumeRole()` ??

Reply:
Good question...